src/main/java/com/google/devtools/build/lib/runtime/CountingArtifactGroupNamer.java

```diff
@@ -13,41 +13,76 @@
 // limitations under the License.
 package com.google.devtools.build.lib.runtime;
 
+import com.google.common.util.concurrent.Uninterruptibles;
 import com.google.devtools.build.lib.buildeventstream.ArtifactGroupNamer;
 import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos.BuildEventId.NamedSetOfFilesId;
 import com.google.devtools.build.lib.collect.nestedset.NestedSet;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CountDownLatch;
 import javax.annotation.concurrent.ThreadSafe;
 
 /** Conversion of paths to URIs. */
 @ThreadSafe
 public class CountingArtifactGroupNamer implements ArtifactGroupNamer {
 
-  private final Map<NestedSet.Node, Integer> nodeNames = new HashMap<>();
+  private final ConcurrentMap<NestedSet.Node, LatchedGroupName> nodeNames =
+      new ConcurrentHashMap<>();
 
   @Override
   public NamedSetOfFilesId apply(NestedSet.Node id) {
-    Integer name;
-    synchronized (this) {
-      name = nodeNames.get(id);
-    }
+    LatchedGroupName name = nodeNames.get(id);
     if (name == null) {
       return null;
     }
-    return NamedSetOfFilesId.newBuilder().setId(name.toString()).build();
+    return NamedSetOfFilesId.newBuilder().setId(name.getName()).build();
   }
 
   /**
    * If the {@link NestedSet} has no name already, return a new name for it. Return null otherwise.
    */
-  public synchronized String maybeName(NestedSet<?> set) {
+  public LatchedGroupName maybeName(NestedSet<?> set) {
     NestedSet.Node id = set.toNode();
-    if (nodeNames.containsKey(id)) {
+    LatchedGroupName existingGroupName;
+    LatchedGroupName newGroupName;
+    // synchronized necessary only to ensure node names are chosen uniquely and compactly.
+    // TODO(adgar): consider dropping compactness and unconditionally increment an AtomicLong to
+    // pick unique node names.
+    synchronized (this) {
+      newGroupName = new LatchedGroupName(nodeNames.size());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *newGroupName*. The resource is referenced at lines 53, 55, 56. The resource is returned at line 59. There are other execution paths that don&#39;t close the resource or return it.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).